backend/src/apiserver/storage/run_store.go

```diff
@@ -152,7 +152,10 @@ func (s *RunStore) buildSelectRunsQuery(selectCount bool, opts *list.Options,
 	if refKey != nil && refKey.Type == "ExperimentUUID" {
 		// for performance reasons need to special treat experiment ID filter on runs
 		// currently only the run table have experiment UUID column
-		filteredSelectBuilder, err = list.FilterRunOnExperiment("run_details", runColumns,
+		filteredSelectBuilder, err = list.FilterOnExperiment("run_details", runColumns,
+			selectCount, refKey.ID)
+	} else if refKey != nil && refKey.Type == common.Namespace {
+		filteredSelectBuilder, err = list.FilterOnNamespace("run_details", runColumns,
```

Comment:
Discussion:
I think the comment above mentioned this is a special optimization for performance reason, should we add this adhoc logic?
Alternative is to keep using resource reference for filtering like below, for that to work, we need to add namespace resource reference if it doesn't exist.

I'm not familiar with these, just want to know perf/cleanness tradeoffs. What did you have in mind when implementing this?
Probably, @IronPan for review too

Reply:
and should UI keep adding namespace resource reference to create run api requests or not?

Reply:
Please correct me if i'm wrong. The goal here is for two pages - 1. list runs for a namespace, 2. list all the runs.
In 2nd case, we still want the behavior to be filtering on the namespace user selected.

Is that correct?

Reply:
I'm not sure I understood you,
here's what I thought, two goals:
1. list by a specific experiment
2. list by namespace

Reply:
and I want to ask if we keep using previous convention of resource reference to do `list by namespace` or we introduce adhoc filtering code like here.

Reply:
@Bobgy 

> I think the comment above mentioned this is a special optimization for performance reason, should we add this adhoc logic?
> Alternative is to keep using resource reference for filtering like below, for that to work, we need to add namespace resource reference if it doesn't exist.
> 
The optimization is for filtering by experiment only. It was a recent change made by some community developer, a new column `ExperimentUUID` was added to the `run_details` table. So if the filter type is `ExperimentUUID`, we query directly from `run_details` table without any external join, whereas if the filter type is `Experiment`, we fall back to query from two tables: `run_details` and `resource_references`. We need to have both logic for backward compatibility, as old runs created before this change doesn't have the `ExperimentUUID` value. And this PR doesn't change this behavior. I mentioned this briefly that we should apply the same change to jobs as well (a P2 item).

For filtering by namespace, the `Namespace` column exists in `run_details` since the very beginning, so we can use this as the single source of truth. As discussed in the design doc, its value could differ from the namespace from resource_references, and we will rely on the sql migration script to resolve any possible difference.

> and should UI keep adding namespace resource reference to run api requests or not?

That's a good question, I think UI can stop adding namespace resource reference to `CreateRunRequest`, if backend receives namespace in resource reference, it will still store it in `resource_references` table, though the value will never be read once this change merged. So we should clean up these code later.

Reply:
@IronPan

> Please correct me if i'm wrong. The goal here is for two pages - 1. list runs for a namespace, 2. list all the runs.
> In 2nd case, we still want the behavior to be filtering on the namespace user selected.
> 
> Is that correct?

Yes, that's correct.

Reply:
I see, so we do it for perf optimization. I just wanted to double check why we do it. SGTM

Does it worth adding comment like above case too?
> // for performance reasons need to special treat experiment ID filter on runs

Reply:
> Does it worth adding comment like above case too?
> 

You mean we add a similar comment for the case when `refKey.Type == common.Namespace`? 
I think we should be fine without the comment since the code path of retrieving namespace from `resource_references` table was newly introduced when we started the multi-user effort. So it's like "nonexistent" to almost all users. And we will deprecate this code path shortly.